File: rust/cudf-sys/cpp/quantile.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#include "cudf-sys/src/quantile.rs.h"
#include "cudf-sys/src/lib.rs.h"
#include "cudf-sys/helpers.hpp"

#include <cudf/quantiles.hpp>
#include <cudf/tdigest/tdigest_column_view.hpp>

namespace cudf_sys {

std::unique_ptr<Column> quantile_column(
    cudf::column_view const& col,
    rust::Slice<double const> quantiles,
    int32_t interp,
    std::size_t stream) {
  std::vector<double> q(quantiles.begin(), quantiles.end());
  return COL(cudf::quantile(col, q, ENUM<cudf::interpolation>(interp), cudf::column_view{}, true, S(stream)));
}

std::unique_ptr<Table> quantiles_table(
    Table const& tbl,
    rust::Slice<double const> quantiles,
    int32_t interp,
    bool is_input_sorted,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  std::vector<double> q(quantiles.begin(), quantiles.end());
  auto sorted = is_input_sorted ? cudf::sorted::YES : cudf::sorted::NO;
  return TBL(cudf::quantiles(tbl.cached_view(), q,
      ENUM<cudf::interpolation>(interp), sorted,
      ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Column> percentile_approx(cudf::column_view const& tdigest_col, cudf::column_view const& percentiles, std::size_t stream) {
  return COL(cudf::percentile_approx(cudf::tdigest::tdigest_column_view(tdigest_col), percentiles, S(stream)));
}

}  // namespace cudf_sys

File: rust/cudf-sys/src/quantile.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#[cxx::bridge(namespace = "cudf_sys")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cudf-sys/lib.hpp");

        // Shared types from the core bridge.
        type Column = crate::ffi::Column;
        type Table = crate::ffi::Table;
        #[namespace = "cudf"]
        type column_view = crate::ffi::column_view;

        // -- Quantile operations --

        /// Computes quantiles of a column.
        fn quantile_column(
            col: &column_view,
            quantiles: &[f64],
            interp: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        // -- Quantiles (new) --

        /// Table-level quantile rows.
        fn quantiles_table(
            tbl: &Table,
            quantiles: &[f64],
            interp: i32,
            is_input_sorted: bool,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        // -- Percentile approx --

        /// Computes approximate percentiles from a t-digest column.
        fn percentile_approx(
            tdigest_col: &column_view,
            percentiles: &column_view,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;
    }
}
